software/tetris/main.c

```diff
@@ -0,0 +1,999 @@
+    scr_putc(BOARD_Y + r, BOARD_X + COLS, 0, CHAR_FLOOR_RIGHT);
+}
+
+/**
+ * Display the current level and score values, and paused state.
+ */
+static void display_score(void) {
+    const uint16_t blink = PAL_BLINK << SCROLL_MAP_PAL_SHIFT;
+
+    scr_puts(INFO_Y, INFO_X + 0, 0, "Level: ");
+    scr_zpadnum(INFO_Y, INFO_X + 7, 0, level, 2);
+
+    scr_puts(INFO_Y + 1, INFO_X + 0, 0, "Score: ");
+    scr_zpadnum(INFO_Y + 1, INFO_X + 7, 0, score, 4);
+
+    if (state & PAUSED) {
+        scr_puts(INFO_Y + 3, INFO_X + 0, blink, "[Paused]");
+    } else {
+        scr_puts(INFO_Y + 3, INFO_X + 0, blink, "        ");
+    }
+}
+
+/**
+ * Check for completed rows and remove them from the gaming board.
+ */
+static void check_remove_completed_rows(void) {
+    uint8_t r, removed;
+
+    removed = 0;
+    for (r = 0; r < ROWS; r++) {
+        if (is_complete_row(r)) {
+            removed++;
+            break;
+        }
+    }
+
+    removed = 0;
+    for (r = 0; r < ROWS; r++) {
+        if (is_complete_row(r)) {
+            removed++;
+            remove_row(r);
+            score += SCORE_PER_ROW;
+            if (++lines == 4) {
+                lines = 0;
+                if (level < MAX_level) {
+                    level++;
+                    step_ms = STEP_FASTER(step_ms); // 3/4 times game speedup
+                }
+            }
+        }
+    }
+
+    if (current_row < free_rows)
+        free_rows = current_row;
+    free_rows += removed;
+
+    if (removed) {
+        scr_beep();
+    }
+    display_board();
+    display_score();
+}
+
+/**
+ * Try to move the current block left.
+ * This function updates the current_col variable.
+ */
+static void cmd_move_left(void) {
+  current_col--;
+  if (test_if_block_fits()) {
+      current_xanim = 8;
+      return;
+  }
+
+  // If we arrive here, the block doesn't fit,
+  // and we simply undo the column change.
+  current_col++;
+}
+
+/**
+ * Try to move the current block right.
+ * This function updates the current_col variable.
+ */
+static void cmd_move_right(void) {
+    current_col++;
+    if (test_if_block_fits()) {
+        current_xanim = -8;
+        return;
+    }
+
+    // If we arrive here, the block doesn't fit,
+    // and we simply undo the column change.
+    current_col--;
+}
+
+/**
+ * Try to rotate the current block.
+ * This function updates the current_block variable.
+ */
+static void cmd_rotate(char r) {
+    rotate_block(r);
+    if (test_if_block_fits()) return;
+
+    // If we arrive here, the block doesn't fit,
+    // and we undo the rotation by three more rotations...
+    rotate_block(-r);
+}
+
+// Efficiently center fixed-length message.
+#define CENTER(y, color, message) scr_puts(y, TILES_W / 2 - (sizeof(message) - 1) / 2, color, message)
+
+/**
+ * Called when the game starts initially.
+ */
+static void do_beginning(void) {
+    const uint16_t blink = PAL_BLINK << SCROLL_MAP_PAL_SHIFT;
+    scr_clear();
```

Comment:
To fix that issue I mentioned this `scr_clear()` call could move out of here and into `main()` before the loop. It looks like the only place where it transitions to `BEGINNING` state:
```C
scr_clear();
state = BEGINNING;
```